### PR TITLE
feat(errors): default error serializer content negotiation

### LINKED ISSUE #2349: Add content_type to Result objects returned by testing client

### PR DESCRIPTION
**Pull request by copalco**
# Summary of Changes

We would like to use choose from available content type handlers to return error response content in a best format basing on the accept header of the request.

# Related Issues

Fixes https://github.com/falconry/falcon/issues/2023
Fixes #2349

# Pull Request Checklist

This is just a reminder about the most common mistakes.  Please make sure that you tick all *appropriate* boxes.  But please read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!

If an item doesn't apply to your pull request, **check it anyway** to make it apparent that there's nothing to do.

- [ ] Applied changes to both WSGI and ASGI code paths and interfaces (where applicable).
- [ ] Added **tests** for changed code.
- [ ] Prefixed code comments with GitHub nick and an appropriate prefix.
- [ ] Coding style is consistent with the rest of the framework.
- [ ] Updated **documentation** for changed code.
    - [ ] Added docstrings for any new classes, functions, or modules.
    - [ ] Updated docstrings for any modifications to existing code.
    - [ ] Updated both WSGI and ASGI docs (where applicable).
    - [ ] Added references to new classes, functions, or modules to the relevant RST file under `docs/`.
    - [ ] Updated all relevant supporting documentation files under `docs/`.
    - [ ] A copyright notice is included at the top of any new modules (using your own name or the name of your organization).
    - [ ] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/usage/restructuredtext/directives.html?highlight=versionadded#directive-versionadded).
- [ ] Changes (and possible deprecations) have [towncrier](https://towncrier.readthedocs.io/en/latest/quickstart.html#creating-news-fragments) news fragments under `docs/_newsfragments/`, with the file name format `{issue_number}.{fragment_type}.rst`. (Run `towncrier --draft` to ensure it renders correctly.)

If you have *any* questions to *any* of the points above, just **submit and ask**! This checklist is here to *help* you, not to deter you from contributing!

*PR template inspired by the attrs project.*

### CODE REVIEW

**Comment by CaselIT** [comment]
@copalco I've taken the liberty of improving the implementation.

@vytas7 should be ready for review. It should be mostly backward compatible, but I feel it's still better to have this for 4.0 instead of 4.x

File: docs/_newsfragments/2023.newandimproved.rst [hunk 1]
```diff
@@ -0,0 +1,4 @@
+The default error serializer will now use the response media handlers
+to better negotiate the response format with the client.
```
Comment (CaselIT) [inline docs/_newsfragments/2023.newandimproved.rst:2]: format -> serialization

File: falcon/testing/client.py [hunk 2]
```diff
@@ -274,6 +274,11 @@ def encoding(self) -> Optional[str]:
         """
         return self._encoding
 
+    @property
+    def content_type(self) -> Optional[str]:
```
Comment (CaselIT) [inline falcon/testing/client.py:278]: maybe I should add a specific test for this that makes use also of the stream response

**Review by CaselIT — APPROVED** [review]

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thanks for the effort @CaselIT and @copalco!

I agree it would be nice to squeeze it into a major release since it is indeed bordering a breaking change.

I'll review the implementation more in detail later, but unfortunately this needs a bit of design rework as well as it fails my testing with a generic `async` handler.

**Comment by vytas7** [comment]
This is a simple test app that I used:
```python
import falcon.asgi
import falcon.media


class AsyncHandler(falcon.media.BaseHandler):
    async def serialize_async(self, media, content_type):
        return f'{media}\n'.encode()


class Resource:
    async def on_get(self, req, resp):
        if not req.get_param('password'):
            raise falcon.HTTPForbidden(title='NotAllowed')

        resp.content_type = 'falcon/peregrine'
        resp.media = {'msg': 'Hello'}


app = falcon.asgi.App()
app.resp_options.media_handlers['falcon/peregrine'] = AsyncHandler()
app.add_route('/', Resource())
```

Rendering normal media using my new handler works:
```
$ xh http://localhost:8000/?password=secret Accept:falcon/peregrine

HTTP/1.1 200 OK
Content-Length: 17
Content-Type: falcon/peregrine
Date: Sun, 29 Sep 2024 10:44:42 GMT
Server: uvicorn

{'msg': 'Hello'}
```

Going via the exception route fails with an internal server error:
```
$ xh http://localhost:8000/ Accept:falcon/peregrine

HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 21
Content-Type: text/plain; charset=utf-8
Date: Sun, 29 Sep 2024 10:45:39 GMT
Server: uvicorn

Internal Server Error
```

```
Traceback (most recent call last):
  File "/home/vytas/.virtualenvs/falcon-py312/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 401, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vytas/.virtualenvs/falcon-py312/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vytas/progs/tmp/copalco/falcon/falcon/asgi/app.py", line 529, in __call__
    if not await self._handle_exception(req, resp, ex, params):
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vytas/progs/tmp/copalco/falcon/falcon/asgi/app.py", line 1299, in _handle_exception
    await err_handler(req, resp, ex, params, **kwargs)
  File "/home/vytas/progs/tmp/copalco/falcon/falcon/asgi/app.py", line 1207, in _http_error_handler
    self._compose_error_response(req, resp, error)
  File "/home/vytas/progs/tmp/copalco/falcon/falcon/app.py", line 1126, in _compose_error_response
    self._serialize_error(req, resp, error)
  File "/home/vytas/progs/tmp/copalco/falcon/falcon/app_helpers.py", line 328, in default_serialize_error
    resp.data = handler.serialize(exception.to_dict(), preferred)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/vytas/progs/tmp/copalco/falcon/falcon/media/base.py", line 61, in serialize
    raise NotImplementedError()
NotImplementedError
```

**Comment by CaselIT** [comment]
good catch! fixed it

File: falcon/app_helpers.py [hunk 3]
```diff
@@ -311,13 +310,27 @@ def default_serialize_error(req: Request, resp: Response, exception: HTTPError)
...
         if preferred == MEDIA_JSON:
-            handler, _, _ = resp.options.media_handlers._resolve(
-                MEDIA_JSON, MEDIA_JSON, raise_not_found=False
-            )
+            # NOTE(caselit): special case json to ensure that it's always possible to
+            # serialize an error in json even if no handler is set in the
+            # media_handlers.
             resp.data = exception.to_json(handler)
+        elif handler:
+            if serialize_sync:
```
Comment (CaselIT) [inline falcon/app_helpers.py:328]: @vytas7 not sure we need to bother with this optimization here.

Comment (vytas7) [reply falcon/app_helpers.py:328]: I don't think we need this path, there is already the same optimization in place for rendering out the body from media. Just set the media and let the app do this.

Comment (CaselIT) [reply falcon/app_helpers.py:328]: I agree. Will simplify

**Comment by CaselIT** [comment]
@vytas7 updated

File: falcon/app_helpers.py [hunk 4]
```diff
@@ -311,13 +310,25 @@ def default_serialize_error(req: Request, resp: Response, exception: HTTPError)
...
+        )
         if preferred == MEDIA_JSON:
-            handler, _, _ = resp.options.media_handlers._resolve(
-                MEDIA_JSON, MEDIA_JSON, raise_not_found=False
-            )
+            # NOTE(caselit): special case json to ensure that it's always possible to
+            # serialize an error in json even if no handler is set in the
+            # media_handlers.
             resp.data = exception.to_json(handler)
+        elif handler:
```
Comment (vytas7) [inline falcon/app_helpers.py:327]: As I understand from the code, adding an XML media handler should take precedence over `.to_xml()`.
Could we maybe add a unit test for this case?

**Review by vytas7 — CHANGES_REQUESTED** [review]
Thanks, looks good bar some minor remarks (commented inline).

We also need to make a decision on being able to disable XML, and whether to make a breaking change, and disable XML by default.

File: docs/_newsfragments/content_type.newandimproved.rst
[the remark this answers is not in the document]
Comment (CaselIT) [reply docs/_newsfragments/content_type.newandimproved.rst:1]: will use the PR number. It seemed useful so added it. Can add an issue maybe?

File: falcon/app_helpers.py [hunk 5]
```diff
@@ -311,13 +310,25 @@ def default_serialize_error(req: Request, resp: Response, exception: HTTPError)
             preferred = MEDIA_JSON
         elif '+xml' in accept:
             preferred = MEDIA_XML
+        else:
+            # NOTE(caselit): if nothing else matchers try using the media handlers
+            # registered in the response
+            preferred = req.client_prefers(list(resp.options.media_handlers))
```
[the remark this answers is not in the document]
Comment (CaselIT) [reply falcon/app_helpers.py:316]: We can update the typing of client prefers

File: docs/_newsfragments/content_type.newandimproved.rst
[the remark this answers is not in the document]
Comment (CaselIT) [reply docs/_newsfragments/content_type.newandimproved.rst:1]: #2349

**Comment by CaselIT** [comment]
Updated.

> We also need to make a decision on being able to disable XML, and whether to make a breaking change, and disable XML by default.

Not too sure about this. Let's merge this one and think about it after

**Comment by vytas7** [comment]
As discussed on Gitter, we don't want to make a breaking change wrt XML right now.

**Review by vytas7 — APPROVED** [review]
LGTM now, thanks @copalco and @CaselIT!

**Pull request merged**